### PR TITLE
feat(users): implement GET single user API

### PR DESCRIPTION
**Pull request by Sukhacoder02**
- implemented GET /users/:user_id to get a single user with given ```user_id``` from the db.
Screenshot 2023-02-10 at 11 02 45 AM
- added test cases for user.service
- added test cases for user.controller
Screenshot 2023-02-10 at 10 59 36 AM

### CODE REVIEW

File: src/services/user.service.js [hunk 1]
```diff
@@ -1,10 +1,22 @@
 
 const { users } = require('../models');
+const CustomErrors = require('../../errors/httpError');
 
 const userServices = {
   getAllUsers: async () => {
     const allUsers = await users.findAll();
     return allUsers;
+  },
+  getOneUser: async (user_id) => {
```
Comment (rohini-senthil) [inline src/services/user.service.js:10]: Can we rename these functions to `listUsers` and `getUser`

Comment (Sukhacoder02) [reply src/services/user.service.js:10]: Sure, working on these.

Comment (Sukhacoder02) [reply src/services/user.service.js]: Done, waiting for review.

File: src/controllers/user.controller.js [hunk 2]
```diff
@@ -1,11 +1,27 @@
 
 
 const userServices = require('../services/user.service');
+const { NotFoundError } = require('../../errors/httpError');
 
 const getUsers = async (_, res) => {
```
Comment (rohini-senthil) [inline src/controllers/user.controller.js:6]: Can we rename this to `listUsers` as well?

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -1,11 +1,27 @@
...
-  const allUsers = await userServices.getAllUsers();
+  const allUsers = await userServices.listUsers();
   res.status(200);
   res.json(allUsers);
 };
-
-module.exports = { getUsers };
\ No newline at end of file
+const getUserById = async (req, res) => {
```
Comment (rohini-senthil) [inline src/controllers/user.controller.js:11]: And rename this to just `getUser`?

File: src/controllers/user.controller.js (hunk 2 above)
Comment (Sukhacoder02) [reply src/controllers/user.controller.js:6]: So, should I use the same function names for all controllers as well as services?

Comment (rohini-senthil) [reply src/controllers/user.controller.js:6]: In this case, yes.

Comment (Sukhacoder02) [reply src/controllers/user.controller.js:6]: I'll surely get to that.

File: src/routes/user.route.js [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
 const validator = require('../middlewares/request.validator');
 // require schemas
 const schemas = require('../middlewares/schemas.validator');
 
 const router = express.Router();
 
 router.get('/users', listUsers);
-router.post('/users', validator(schemas.userSchema, 'body'), postUser);
+router.get('/users/:user_id', validator(schemas.userIdSchema, 'params'), getUser);
+// router.post('/users', validator(schemas.userSchema, 'body'), postUser);
```
Comment (rohini-senthil) [inline src/routes/user.route.js:11]: Why is the create user route commented out?

File: src/routes/user.route.js [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
+const { listUsers, postUser, deleteUser, getUser } = require('../controllers/user.controller');
 const validator = require('../middlewares/request.validator');
 // require schemas
 const schemas = require('../middlewares/schemas.validator');
 
 const router = express.Router();
 
 router.get('/users', listUsers);
-router.post('/users', validator(schemas.userSchema, 'body'), postUser);
+router.get('/users/:user_id', validator(schemas.userIdSchema, 'params'), getUser);
```
Comment (rohini-senthil) [inline src/routes/user.route.js:10]: Can you also add auth validation to this?

Comment (Sukhacoder02) [reply src/routes/user.route.js]: Added.

File: src/middlewares/schemas.validator.js [hunk 6]
```diff
@@ -1,5 +1,11 @@
-// require joi
 const Joi = require('joi');
+
+const userIdSchema = Joi.object({
```
Comment (rohini-senthil) [inline src/middlewares/schemas.validator.js:3]: I think we can directly do 
```
Joi.string().uuid({
  version: ['uuidv4', 'uuidv1'],
});
```
Also, let's move user related validators to `user.validator.js`

File: tests/controllers/user.controller.test.js [hunk 7]
```diff
@@ -107,55 +108,78 @@ describe('User Controller', () => {
...
-        message: 'User deleted',
-      });
+      const err = new NotFoundError('User not found');
+      jest.spyOn(userServices, 'getUser').mockRejectedValue(err);
+      await userController.getUser(mockReq, mockRes);
+      expect(mockRes.status).toHaveBeenCalledWith(err.code);
+      expect(mockRes.json).toHaveBeenCalledWith({ message: err.message });
     });
   });
+  // describe('deleteUser test', () => {
```
Comment (rohini-senthil) [inline tests/controllers/user.controller.test.js:158]: Why is this commented?

File: tests/controllers/user.controller.test.js [hunk 8]
```diff
@@ -49,35 +19,66 @@ describe('Check User Controller', () => {
...
+          updatedAt: '2023-02-09T15:02:53.658Z',
+          createdAt: '2023-02-09T15:02:53.658Z',
+          fmno: null,
+          current_engagement_ids: null,
+          case_study_ids: null,
+          skills: null,
+          role: null,
+          guild: null,
+          past_engagement_ids: null,
+          image: null,
```
Comment (rohini-senthil) [inline tests/controllers/user.controller.test.js:53]: Can we move common mocks to a `mocks` folder under tests folder. It's starting to crowd the file

**Review by rohini-senthil — CHANGES_REQUESTED** [review]
Requested changes

File: src/utils/HttpError.js [hunk 9]
```diff
@@ -4,4 +4,11 @@ class HttpError extends Error {
     this.statusCode = code;
```
Comment (rohini-senthil) [inline src/utils/HttpError.js:4]: Can we change the file name to httpError?

Comment (Sukhacoder02) [reply src/utils/HttpError.js]: Done, waiting for review.

Comment (Sukhacoder02) [reply src/utils/HttpError.js]: Oops, still some imports left to change, will try to do so as soon as possible.
Sorry.

Comment (Sukhacoder02) [reply src/utils/HttpError.js]: I've renamed all the imports to camelCase and CI tests are successful.
Waiting for merger.

**Review by rohini-senthil — APPROVED** [review]

**Pull request merged**